rfcs/20190729-keras-preprocessing.md

```diff
@@ -0,0 +1,738 @@
+# Keras Preprocessing API
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Francois Chollet (fchollet@google.com)               |
+| **Updated**   | 2019-07-29                                           |
+
+
+## Context
+
+`tf.data.Dataset` is the main API for data loading and preprocessing in TensorFLow. It has two advantages:
+
+- It supports GPU prefetching
+- It supports distribution via the Distribution Strategies API
+
+Meanwhile, `keras.preprocessing` in a major API for data loading and preprocessing in Keras. It is based
+on Numpy and Scipy, and it produces instances of the `keras.utils.Sequence` class, which are finite-length,
+resettable Python generators that yield batches of data.
+
+Some features of `keras.preprocessing` are highly useful and don't have straightforward equivalents in `tf.data`
+(in particular image data augmentation and dynamic time series iteration).
+
+Ideally, the utilities in `keras.preprocessing` should be made compatible with `tf.data`.
+This presents the opportunity to improve on the existing API. In particular we don't have good support
+for image segmentation use cases today.
+
+Some features are also being subplanted by [preprocessing layers](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md), in particular text processing. 
+As a result we may want to deprecate them.
+
+
+## Goals
+
+- Make all features of `keras.preprocessing` compatible with `tf.data`
+- As a by-product, add required ops to TensorFlow
+- Improve the ergonomy and user-friendliness of the `keras.preprocessing` APIs
+
+
+## Proposed changes at a high-level
+
+- Remove the submobules `image`, `text`, `timeseries` from the public API and expose their contents as part of the top-level `keras.preprocessing`.
+- Make all classes in `keras.preprocessing` JSON-serializable.
+- Deprecate `Tokenizer` class in favor of `TextVectorization` preprocessing layer
+- Make image-transformation functions in `affine_transformations` submodule compatible with TensorFlow tensor inputs (accepting tensors and returning tensors).
+- Improve signature of the above-mentioned functions, by using fully-spelled argument names, and fewer arguments if possible.
+- Add `dataset_from_arrays`, `dataset_from_dataframe`, and `dataset_from_directory` methods in `ImageDataGenerator` (note: we could use `*_to_*` as well).
+- Rename methods `flow` to `generator_from_arrays`, `flow_from_directory` to `generator_from_directory`, and `flow_from_dataframe` to `generator_to_dataframe`.
+- Improve signature of the above-mentioned methods.
+- Figure out how to support image segmentation use cases with `ImageDataGenerator` (open question).
+- Refactor `TimeseriesGenerator` to follow a similar design as `ImageDataGenerator`. Only configuration arguments should be passed in the constructor. The data should be passed to methods such as `dataset_from_arrays`.
```

Comment:
I don't think people reuse their ImageDataGenerator, but both calls to `ImageDataGenerator` and `dataset_from_dataframe` requires many arguments which might confuse the user.

Reply:
The factoring between config and data is there for 2 reasons:
- to enable reusing a config on new data; this is how we plan to support image segmentation for instance (target masks have to be augmented in the same way as inputs)
- makes it  easier for people to think about data augmentation/normalization and loading separately, in two steps, which would otherwise form a very large and potentially confusing single step.